### PR TITLE
Adding depth-wise table stacking 

### PR DESCRIPTION
**Pull request by himanshupathak21061998**
Trying to fixing:#6660
Not ready for a final review but contains an implementation of cstack method for depth wise stacking.

### CODE REVIEW

**Comment by taldcroft** [comment]
@himanshupathak21061998 - thanks, looks like a good start.  Indeed cstack is quite similar to vstack, and going back #6660 the crux was about `new_like`.  Your implementation ends up passing a tuple as `length`, which happens to work for `Column`, but in general `length` is currently required by API to be an `int`.  For instance if you try to `cstack` with `Time` objects it fails.  So either we need to do something with `new_like`, or something else...

Looking at this I got an idea to make this implementation super simple.  Roughly speaking, cstack is just exactly vstack, but with the final output columns reshaped correctly.  So I think it might work to literally call `vstack`, get that table, and then reshape and transpose the output columns.

One hint, which is a brand new feature of table.  You can avoid making a new output table object with the following:
```
    # check tables all same length
    out = vstack(tables, join_type, metadata_conflicts)

    for name, col in out.columns.items():
        new_col = reshaped_and_transposed(col)
        out.columns.__setitem__(name, new_col, validated=True)
```

The reshape and transpose is simple for input 1-d columns, but getting it straight for input N-d columns is a little trick and that is a challenge left for you!  :smile:

**Comment by himanshupathak21061998** [comment]
Thanks for helping @taldcroft and sorry for the delay. Changes updated now I will start writing tests for this.

**Comment by bsipocz** [comment]
@himanshupathak21061998 - if you rebase, the 32bit errors would go away.

**Comment by taldcroft** [comment]
@himanshupathak21061998 - I think the whole thing can be done far more simply by fully-using the similarity of vstack and cstack.  I thought you might get this independently, but here is a solution that seems to work:
```
def cstack(tables, join_type='outer', metadata_conflicts='warn'):
    n_rows = set(len(table) for table in tables)
    if len(n_rows) != 1:
        raise ValueError()
    n_row = n_rows.pop()

    out = vstack(tables, join_type, metadata_conflicts)

    for name, col in out.columns.items():
        col = out[name]

        # Reshape to so each original column is now in a row.
        # If entries are not 0-dim then those additional shape dims
        # are just carried along.
        # [x x x y y y] => [[x x x],
        #                   [y y y]]
        col.shape = (len(tables), n_row) + col.shape[1:]

        # Transpose the table and row axes to get to
        # [[x, y],
        #  [x, y]
        #  [x, y]]
        axes = np.arange(len(col.shape))
        axes[:2] = [1, 0]

        # This temporarily makes `out` be corrupted (columns of different
        # length) but it all works out in the end.
        out.columns.__setitem__(name, col.transpose(axes), validated=True)

    return out
```
BTW, no need to follow the legacy pattern of having an internal `_stack` method.  That is from the ancient days when we supported stacking numpy structured arrays.

**Comment by taldcroft** [comment]
Example:
```
In [15]: t = simple_table()

In [16]: t2 = simple_table()

In [17]: c = cstack([t, t2])

In [18]: c2 = cstack([t, t2])

In [19]: d = cstack([c, c2])

In [20]: d['a'].shape
Out[20]: (3, 2, 2)

In [21]: c['a'].shape
Out[21]: (3, 2)

In [22]: t['a'].shape
Out[22]: (3,)

In [23]: d['a'][0]
Out[23]: 
array([[1, 1],
       [1, 1]])

In [24]: e = cstack([t, c])
---------------------------------------------------------------------------
TableMergeError                           Traceback (most recent call last)
<ipython-input-24-71e13d610856> in <module>()
----> 1 e = cstack([t, c])
...
TableMergeError: columns have different shapes

```

**Comment by himanshupathak21061998** [comment]
> In [20]: d['a'].shape
> Out[20]: (3, 2, 2)

Sorry, but confusion was because of this case I thought it be like this
```
d['a'].shape
Out[4]: (3, 4)
```
My mistake I thought that shape with more than two elements tuple is not allowed. Next time I will try better thanks @taldcroft

File: astropy/table/tests/test_operations.py [hunk 1]
```diff
@@ -961,6 +961,105 @@ def test_mixin_functionality(self, mixin_cols):
...
+        assert type(t35['a']) is type(t3['a'])
+        assert type(t35['b']) is type(t3['b'])
+        assert t35.pformat() == [' a [2,2]    b [2,2]  ',
+                                 '---------- ----------',
+                                 '5.0 .. 2.0 foo .. pez',
+                                 '4.0 .. 6.0 pez .. sez']
+
+        with pytest.raises(TableMergeError):
+            table.cstack([t2, t3])
+
```
Comment (himanshupathak21061998) [inline astropy/table/tests/test_operations.py]: Is it okay that we cannot merge column having a different shape for cstack?

File: astropy/table/operations.py [hunk 2]
```diff
@@ -231,6 +232,88 @@ def setdiff(table1, table2, keys=None):
...
+      --- ---
+        5   7
+        6   8
+      >>> print(cstack([t1, t2]))
+      a [2]  b [2]
+      ------ ------
+      1 .. 5 3 .. 7
+      2 .. 6 4 .. 8
+    """
+    n_rows = set(len(table) for table in tables)
```
Comment (himanshupathak21061998) [inline astropy/table/operations.py]: If we enter `cstack([t1, t1[1]])` it returns true should I change this behaviour. I thought it is better to ask before doing?

**Comment by himanshupathak21061998** [comment]
@taldcroft  thanks for helping. I have performed all changes as required.

File: astropy/table/tests/test_operations.py [hunk 3]
```diff
@@ -961,6 +961,105 @@ def test_mixin_functionality(self, mixin_cols):
...
+        self.t2 = t_cls.read([' a    b   c',
+                              ' 2.  pez  4',
+                              ' 3.  sez  5'], format='ascii')
+
+        self.t3 = t_cls({'a': [[5., 6.], [4., 3.]],
+                         'b': [['foo', 'bar'], ['pez', 'sez']]},
+                        names=('a', 'b'))
+        self.t4 = t_cls(self.t1, copy=True, masked=t_cls is Table)
+        self.t5 = t_cls({'a': [[4., 2.], [1., 6.]],
```
Comment (taldcroft) [inline astropy/table/tests/test_operations.py]: One of the inputs should have a masked column with a masked element.

File: astropy/table/tests/test_operations.py [hunk 4]
```diff
@@ -961,6 +961,105 @@ def test_mixin_functionality(self, mixin_cols):
...
+                                 '1.0 .. 1.0 bar .. --']
+
+    def test_stack_basic_outer(self, operation_table_type):
+            if operation_table_type is QTable:
+                pytest.xfail('Quantity columns do not support masking.')
+            self._setup(operation_table_type)
+            t1 = self.t1
+            t2 = self.t2
+            t4 = self.t4
+            t12 = table.cstack([t1, t2], join_type='outer')
```
Comment (taldcroft) [inline astropy/table/tests/test_operations.py]: The `t12` and `t124` tests are basically the same and not improving coverage.  So just keep the `t124` test.

File: astropy/table/tests/test_operations.py [hunk 5]
```diff
@@ -961,6 +961,105 @@ def test_mixin_functionality(self, mixin_cols):
...
+                                      '0.0 .. 0.0 foo .. foo -- .. --',
+                                      '1.0 .. 1.0 bar .. bar -- .. --']
+
+    def test_stack_basic_inner(self, operation_table_type):
+        self._setup(operation_table_type)
+        t1 = self.t1
+        t2 = self.t2
+        t4 = self.t4
+
+        t12 = table.cstack([t1, t2], join_type='inner')
```
Comment (taldcroft) [inline astropy/table/tests/test_operations.py]: Same comment as before, delete the `t12` test.

File: astropy/table/tests/test_operations.py (hunk 1 above)
Comment (taldcroft) [reply astropy/table/tests/test_operations.py]: Yes. This sort of stacking depends on the input column shapes being the same. The same happens for numpy vstack:
```
In [16]: a = np.array([[1, 2]])

In [17]: b = np.array([[3, 4]])

In [18]: np.vstack([a, b])
Out[18]: 
array([[1, 2],
       [3, 4]])

In [19]: b = np.array([[[3, 33], [4, 44]]])

In [20]: np.vstack([a, b])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/git/astropy/cstack.py in <module>()
----> 1 np.vstack([a, b])

~/miniconda3/envs/astropy/lib/python3.6/site-packages/numpy/core/shape_base.py in vstack(tup)
    232 
    233     """
--> 234     return _nx.concatenate([atleast_2d(_m) for _m in tup], 0)
    235 
    236 def hstack(tup):

ValueError: all the input arrays must have same number of dimensions
```

File: astropy/table/operations.py (hunk 2 above)
Comment (taldcroft) [reply astropy/table/operations.py]: I need to think about this.

File: astropy/table/tests/test_operations.py (hunk 5 above)
Comment (himanshupathak21061998) [reply astropy/table/tests/test_operations.py]: Done

File: astropy/table/tests/test_operations.py (hunk 4 above)
Comment (himanshupathak21061998) [reply astropy/table/tests/test_operations.py]: removed

File: astropy/table/tests/test_operations.py (hunk 3 above)
Comment (himanshupathak21061998) [reply astropy/table/tests/test_operations.py]: added

File: astropy/table/operations.py [hunk 6]
```diff
@@ -231,6 +232,88 @@ def setdiff(table1, table2, keys=None):
...
+    if len(n_rows) != 1:
+        raise ValueError('Tables length are not matching. '
```
[the remark this answers is not in the document]
Comment (himanshupathak21061998) [reply astropy/table/operations.py]: Changed

**Comment by himanshupathak21061998** [comment]
@taldcroft Do I need to include more tests for it.

**Comment by taldcroft** [comment]
@himanshupathak21061998 - I did final review and added a few commits to try to finish this up.  Team effort!

**Comment by bsipocz** [comment]
@taldcroft - I suppose a rebase would solve at least some of the CI issues.

**Comment by taldcroft** [comment]
@bsipocz - I rebased on latest master but it is still failing to build in Travis.  Any clue?

**Comment by bsipocz** [comment]
I see the same issue on master, but don't yet see the cause for it. It's definitely unrelated to this PR.

**Comment by bsipocz** [comment]
@taldcroft - Travis is all green now.

**Pull request merged**